### PR TITLE
Add on_auth_start and on_auth_continue

### PR DESCRIPTION
**Pull request by flxo**
Extended authentification needs callback for `AUTH` packets. Extend the plugin trait and add `on_auth_start` and `on_auth_continue`.

### CODE REVIEW

**Review by TotalKrill — COMMENTED** [review]
Looks nice, could I maybe ask you to change or clarifiy who sends what. Using the "hello" in both client and server is not the most clear :)

File: examples/auth.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    fn acl_check(
+        &mut self,
+        client: &dyn MosquittoClientContext,
+        _acl: AclCheckAccessLevel,
+        _msg: MosquittoMessage,
+    ) -> Result<Success, Error> {
+        println!("Plugin on_acl_check, Client {:?}", client.get_id());
+        let client_id = client.get_id().ok_or(Error::Inval)?;
+
+        if self.authenticated.contains(&client_id) {
```
Comment (TotalKrill) [inline examples/auth.rs]: Is this really necessary? Would not authentication not allow the client to connect even? therefore prohibiting any sub/pub or similar?

This might spark wrongful useage if that is the case, tricking people that they must collect all authenticated clients?

Comment (flxo) [reply examples/auth.rs]: My intention is that a client that a client without client_id cannot be authenticated at all.  I'll update the example

**Comment by flxo** [comment]
Should be complete. I checked with a test client (Hive).

File: examples/auth.rs (hunk 1 above)
Comment (TotalKrill) [reply examples/auth.rs]: I did not know that clients could connect without a client_id. 

If this is because the client_id is optional in the trait mosquittoclientcontext, that is because before auth, the client is not guaranteed to have a client_id, but after auth, it must be there. At least from what I have gathered. 

Basically, I believe that the code you now have, is equal to always ` Ok(Success)`, but with extra capacity for bugs (since you can forget to remove/add to the list etc)

**Review by TotalKrill — COMMENTED** [review]
Okay, I have checked it. 

No problems at all with the implementation, from what I can see. The examples are however sending the wrong usage message (partly my fault as well, )

But the examples are basically showing two things, basic authentication and extended authentication. 

None of the examples actually performs any access checking (authorization) on messages. So before merging, I would like the following changes.

- change acl.rs -> basic-auth.rs and auth.rs -> extended-auth.rs
- either explain how you can get the on_message function to fail somehow, OR remove the logic ( I think you cannot trigger the `Err(Error::Auth)` case except by a  bug in either mosquitto,or this plugin example )  


Basically, someone who wants to handle regular passwords as authentication logic should check basic-auth, and someone who wants challenge-response kind of authentication logic should check extended auth, and it should be a bit more clear from the examples. 

And I do not like to have examples who shows unnecessary usage of the `on_message`, looking at the on_message in the extended-auth example, one could get the impression that handling and remembering authenticated clients MUST be done by the plugin, which I think is misleading.

**Comment by flxo** [comment]
Thanks for reviewing this!

> No problems at all with the implementation, from what I can see. The examples are however sending the wrong usage message (partly my fault as well, )
> 
> But the examples are basically showing two things, basic authentication and extended authentication.
> 
> None of the examples actually performs any access checking (authorization) on messages. So before merging, I would like the following changes.
> 
> * change acl.rs -> basic-auth.rs and auth.rs -> extended-auth.rs

Much more cleaner. Done.

> * either explain how you can get the on_message function to fail somehow, OR remove the logic ( I think you cannot trigger the `Err(Error::Auth)` case except by a  bug in either mosquitto,or this plugin example )

> And I do not like to have examples who shows unnecessary usage of the `on_message`, looking at the on_message in the extended-auth example, one could get the impression that handling and remembering authenticated clients MUST be done by the plugin, which I think is misleading.

I'm confused. There's no impl of `on_message` in the extended auth example. Anyways, I removed the `HashSet` with the list of authenticated clients. It doesn't make sense because any return beside `Err(Error::AuthContinue(..))` or `Ok(Success)` will disconnect the client (this is done by mosquito [here](https://github.com/eclipse/mosquitto/blob/a8448a9c7b14bdaee6ec80419d43fd6544e789b6/src/handle_auth.c#L101-L158) - follow some caller up to `mosquitto__loop_rc_handle`.

**Review by TotalKrill — APPROVED** [review]
Top notch work! 

The comment regarding on_message, was actually referring to acl_check. I got them mixed in my head for some reason, sorry about being confusing about that. 

But you fixed it the way I meant, so everything went OK anyway! Very nice work, thank you!
